QuickFIXn/SocketReader.cs

```diff
@@ -71,46 +75,44 @@ public void Read()
-            // Any changes made here should also be made there.
-            try
-            {
+            try {
                 // Begin read if it is not already started
-                _currentReadRequest ??= _stream.BeginRead(buffer, 0, buffer.Length, null, null);
+                _currentReadTask ??= _stream.ReadAsync(buffer, 0, buffer.Length, _readCancellationTokenSource.Token);
 
                 // Wait for it to complete (given timeout)
-                _currentReadRequest.AsyncWaitHandle.WaitOne(timeoutMilliseconds);
+                _currentReadTask.Wait(timeoutMilliseconds);
 
-                if (_currentReadRequest.IsCompleted)
-                {
-                    // Make sure to set _currentReadRequest to before retreiving result
-                    // so a new read can be started next time even if an exception is thrown
-                    var request = _currentReadRequest;
-                    _currentReadRequest = null;
+                if (_currentReadTask.IsCompleted) {
+                    // Dispose/nullify currentReadTask *before* retreiving .Result.
+                    //   Accessting .Result can throw an exception, so we need to reset currentReadTask
+                    //   first, to set us up for the next read even if an exception is thrown.
+                    Task<int>? request = _currentReadTask;
+                    _currentReadTask.Dispose();
+                    _currentReadTask = null;
 
-                    int bytesRead = _stream.EndRead(request);
+                    int bytesRead = request.Result; // (As mentioned above, this can throw an exception!)
                     if (0 == bytesRead)
-                        throw new SocketException(System.Convert.ToInt32(SocketError.Shutdown));
+                        throw new SocketException(Convert.ToInt32(SocketError.Shutdown));
 
                     return bytesRead;
                 }
 
                 return 0;
             }
-            catch (IOException ex) // Timeout
+            catch (AggregateException ex) // Timeout
             {
-                var inner = ex.InnerException as SocketException;
-                if (inner?.SocketErrorCode == SocketError.TimedOut)
-                {
+                IOException? ioException = ex.InnerException as IOException;
+                SocketException? inner = ioException?.InnerException as SocketException;
+                if (inner is not null && inner.SocketErrorCode == SocketError.TimedOut) {
                     // Nothing read 
                     return 0;
                 }
-                else if (inner is not null)
-                {
+
+                if (inner is not null) {
                     throw inner; //rethrow SocketException part (which we have exception logic for)
                 }
-                else
-                    throw; //rethrow original exception
+
+                throw; //rethrow original exception
```

Comment:
If you care about unwrapping the `AggregateException`, you could do:

```suggestion
                if (ex.InnerExceptions.Count == 1)
                {
                    throw ex.InnerExceptions[0];
                }
                
                throw; //rethrow original exception
```

Reply:
not a bad idea, but I think I'll leave it as-is